File: apps/mobile/src/transport/terminal-multiplex/delivery.ts
import { encodeTerminalMultiplexAckRecord } from '@yiru/runtime-protocol/terminal-multiplex/flow-records'
import {
  TerminalMultiplexOpcode,
  type TerminalMultiplexFrame,
  type TerminalMultiplexOpcode as TerminalMultiplexOpcodeValue
} from '@yiru/runtime-protocol/terminal-multiplex/frame'
import { decodeTerminalMultiplexJson } from '@yiru/runtime-protocol/terminal-multiplex/json'
import { decodeTerminalMultiplexSnapshotEndRecord } from '@yiru/runtime-protocol/terminal-multiplex/snapshot-records'

import { MobileTerminalOrderedEvents } from './ordered-events'
import { MobileTerminalSnapshotAssembler } from './snapshot'
import type { MobileTerminalCallbacks, MobileTerminalSnapshot } from './types'

type SendFrame = (
  opcode: TerminalMultiplexOpcodeValue,
  routeId: number,
  seq: bigint,
  correlationId: number,
  payload?: Uint8Array<ArrayBufferLike>
) => boolean

type PendingOutput = {
  payload: Uint8Array<ArrayBufferLike>
  startSeq: bigint
  endSeq: bigint
}

type MobileTerminalDeliveryOptions = {
  routeId: number
  callbacks: MobileTerminalCallbacks
  send: SendFrame
  allocateCorrelationId: () => number
  setCredit: (bytes: number, reason?: 0 | 1 | 2 | 3) => boolean
  getAckEveryBytes: () => number
  onEnd: () => void
}

export class MobileTerminalDelivery {
  private readonly options: MobileTerminalDeliveryOptions
  private readonly assembler = new MobileTerminalSnapshotAssembler()
  private readonly pendingOutput: PendingOutput[] = []
  private readonly orderedEvents: MobileTerminalOrderedEvents
  private parsedSequence = 0n
  private expectedSequence = 0n
  private snapshotting = true
  private recoveryRequested = false
  private initialSnapshotId = 0
  private awaitingSnapshot: MobileTerminalSnapshot | null = null
  private pendingEndSequence: bigint | null = null

  constructor(options: MobileTerminalDeliveryOptions) {
    this.options = options
    this.orderedEvents = new MobileTerminalOrderedEvents(options.callbacks)
  }

  get parsedSeq(): bigint {
    return this.parsedSequence
  }

  beginInitialSnapshot(snapshotId: number): void {
    this.initialSnapshotId = snapshotId
    this.snapshotting = true
    this.options.setCredit(0)
  }

  beginReveal(): void {
    this.snapshotting = true
    this.options.setCredit(0)
  }

  handle(frame: TerminalMultiplexFrame): boolean {
    if (this.orderedEvents.handle(frame)) {
      return true
    }
    switch (frame.opcode) {
      case TerminalMultiplexOpcode.Output:
        this.handleOutput(frame)
        return true
      case TerminalMultiplexOpcode.SnapshotStart:
        this.snapshotting = this.assembler.start(frame)
        this.recoveryRequested = false
        if (!this.snapshotting) {
          this.recover('invalid snapshot start')
        }
        return true
      case TerminalMultiplexOpcode.SnapshotChunk:
        if (!this.assembler.chunk(frame)) {
          this.recover('snapshot chunk mismatch')
        }
        return true
      case TerminalMultiplexOpcode.SnapshotEnd:
        this.handleSnapshotEnd(frame)
        return true
      case TerminalMultiplexOpcode.End:
        this.pendingEndSequence = frame.seq
        this.publishEndIfParsed()
        return true
      case TerminalMultiplexOpcode.ModelRestore:
        this.handleModelRestore(frame)
        return true
      case TerminalMultiplexOpcode.Epoch:
      case TerminalMultiplexOpcode.Heartbeat:
      case TerminalMultiplexOpcode.Subscribe:
      case TerminalMultiplexOpcode.Subscribed:
      case TerminalMultiplexOpcode.Unsubscribe:
      case TerminalMultiplexOpcode.Error:
      case TerminalMultiplexOpcode.Ack:
      case TerminalMultiplexOpcode.Credit:
      case TerminalMultiplexOpcode.Input:
      case TerminalMultiplexOpcode.Resize:
      case TerminalMultiplexOpcode.Resized:
      case TerminalMultiplexOpcode.ClaimViewport:
      case TerminalMultiplexOpcode.SnapshotRequest:
      case TerminalMultiplexOpcode.VisibilityGate:
      case TerminalMultiplexOpcode.RevealSnapshot:
      case TerminalMultiplexOpcode.Signal:
      case TerminalMultiplexOpcode.Kill:
      case TerminalMultiplexOpcode.FitOverride:
      case TerminalMultiplexOpcode.Driver:
        return false
    }
    return false
  }

  outputParsed(endSeqText: string, receiverQueueBytes = 0): void {
    let endSeq: bigint
    try {
      endSeq = BigInt(endSeqText)
    } catch {
      this.recover('invalid renderer parse sequence')
      return
    }
    if (endSeq <= this.parsedSequence || endSeq > this.expectedSequence) {
      this.recover('renderer parse sequence is outside the delivery window')
      return
    }
    const acknowledgedBytes = endSeq - this.parsedSequence
    if (acknowledgedBytes > 0xffffffffn) {
      this.recover('renderer parse acknowledgement is too large')
      return
    }
    this.parsedSequence = endSeq
    this.orderedEvents.publishThrough(this.parsedSequence)
    this.sendAck(0, 0, endSeq, Number(acknowledgedBytes), receiverQueueBytes)
    this.publishEndIfParsed()
  }

  snapshotParsed(snapshotId: number): void {
    const snapshot = this.awaitingSnapshot
    if (!snapshot || snapshot.id !== snapshotId) {
      this.recover('renderer acknowledged an unknown snapshot')
      return
    }
    const coverageEndSeq = BigInt(snapshot.coverageEndSeq)
    this.awaitingSnapshot = null
    this.parsedSequence = coverageEndSeq
    this.expectedSequence = coverageEndSeq
    this.orderedEvents.publishThrough(this.parsedSequence)
    this.sendAck(2, snapshot.id, coverageEndSeq, 0, 0)
    this.snapshotting = false
    this.options.setCredit(1024 * 1024)
    for (const output of this.pendingOutput.splice(0)) {
      const pendingStart = BigInt(snapshot.pendingDeliveryStartSeq)
      if (output.endSeq > pendingStart && output.endSeq <= coverageEndSeq) {
        continue
      }
      if (output.endSeq <= pendingStart) {
        this.recover('output predates snapshot delivery window')
        return
      }
      this.deliverOutput(output)
    }
    if (snapshot.id === this.initialSnapshotId) {
      this.options.callbacks.onSubscribed?.()
    }
  }

  prepareForNewEpoch(): void {
    this.snapshotting = true
    this.assembler.clear()
    this.pendingOutput.splice(0)
    this.awaitingSnapshot = null
    this.recoveryRequested = false
    this.expectedSequence = this.parsedSequence
    this.orderedEvents.clear()
  }

  private handleOutput(frame: TerminalMultiplexFrame): void {
    const startSeq = frame.seq - BigInt(frame.payload.byteLength)
    if (startSeq < 0n) {
      this.recover('invalid output sequence')
      return
    }
    const output = { payload: frame.payload, startSeq, endSeq: frame.seq }
    if (this.snapshotting) {
      this.pendingOutput.push(output)
    } else {
      this.deliverOutput(output)
    }
  }

  private deliverOutput(output: PendingOutput): void {
    if (output.endSeq <= this.expectedSequence) {
      return
    }
    let payload = output.payload
    let startSeq = output.startSeq
    if (startSeq < this.expectedSequence) {
      payload = payload.slice(Number(this.expectedSequence - startSeq))
      startSeq = this.expectedSequence
    }
    if (startSeq !== this.expectedSequence) {
      this.recover('output sequence gap')
      return
    }
    try {
      const data = new TextDecoder('utf-8', { fatal: true }).decode(payload)
      this.expectedSequence = output.endSeq
      this.options.callbacks.onData(data, {
        endSeq: output.endSeq.toString(),
        wireByteLength: payload.byteLength,
        ackEveryBytes: this.options.getAckEveryBytes()
      })
    } catch {
      this.options.callbacks.onError?.('Mobile terminal output is not valid UTF-8.')
      this.recover('invalid output UTF-8')
    }
  }

  private handleSnapshotEnd(frame: TerminalMultiplexFrame): void {
    const end = decodeTerminalMultiplexSnapshotEndRecord(frame.payload)
    if (!end || end.snapshotId !== frame.correlationId || end.coverageEndSeq !== frame.seq) {
      this.recover('invalid snapshot end')
      return
    }
    if (end.status !== 0) {
      this.assembler.clear()
      this.options.callbacks.onError?.(
        end.status === 2
          ? 'Mobile terminal snapshot exceeded the 2 MiB replay limit.'
          : 'Mobile terminal snapshot is unavailable.'
      )
      return
    }
    const snapshot = this.assembler.end(frame)
    if (!snapshot) {
      this.recover('invalid snapshot')
      return
    }
    this.awaitingSnapshot = snapshot
    this.options.callbacks.onSnapshot(snapshot)
  }

  private handleModelRestore(frame: TerminalMultiplexFrame): void {
    const value = decodeTerminalMultiplexJson(frame.payload)
    if (!value || typeof value.snapshotFollows !== 'boolean') {
      this.recover('invalid model restore record')
      return
    }
    this.snapshotting = true
    this.recoveryRequested = value.snapshotFollows
    this.awaitingSnapshot = null
    this.assembler.clear()
    this.options.setCredit(0)
  }

  private recover(_reason: string): void {
    if (this.recoveryRequested) {
      return
    }
    this.recoveryRequested = true
    this.snapshotting = true
    this.awaitingSnapshot = null
    this.assembler.clear()
    this.options.setCredit(0)
    this.options.send(
      TerminalMultiplexOpcode.SnapshotRequest,
      this.options.routeId,
      this.parsedSequence,
      this.options.allocateCorrelationId(),
      new TextEncoder().encode(JSON.stringify({ requestedScrollbackRows: 1_000 }))
    )
  }

  private sendAck(
    kind: 0 | 1 | 2 | 3,
    correlationId: number,
    seq: bigint,
    bytes: number,
    receiverQueueBytes: number
  ): void {
    this.options.send(
      TerminalMultiplexOpcode.Ack,
      this.options.routeId,
      seq,
      correlationId,
      encodeTerminalMultiplexAckRecord({
        kind,
        status: 0,
        errorCode: 0,
        acknowledgedBytes: bytes,
        cumulativeSeq: seq,
        receiverQueueBytes: Math.max(0, Math.min(0xffffffff, receiverQueueBytes))
      })
    )
  }

  private publishEndIfParsed(): void {
    if (this.pendingEndSequence !== null && this.parsedSequence >= this.pendingEndSequence) {
      this.pendingEndSequence = null
      this.options.callbacks.onEnd?.()
      this.options.onEnd()
    }
  }
}

File: apps/mobile-ios/YiruMobile/Platform/Terminal/Multiplex/Stream/Delivery.swift
import Foundation

nonisolated enum TerminalMultiplexDeliveryError: Error, Equatable, Sendable {
    case invalidOutput
    case invalidAcknowledgement
    case snapshotUnavailable(status: UInt8)
    case repeatedRecovery
}

actor TerminalMultiplexDelivery {
    typealias PublishEvent = @Sendable (TerminalSessionEvent) async -> Void

    private struct PendingOutput: Sendable {
        let bytes: Data
        let startSequence: UInt64
        let endSequence: UInt64
    }

    private let route: TerminalBulkRoute
    private let publishEvent: PublishEvent
    private var assembler = TerminalSnapshotAssembler()
    private var pendingOutput: [PendingOutput] = []
    private var parsedSequence: UInt64 = 0
    private var expectedSequence: UInt64 = 0
    private var initialSnapshotID: UInt32 = 0
    private var awaitingSnapshot: TerminalReplaySnapshot?
    private var pendingEndSequence: UInt64?
    private var isSnapshotting = true
    private var hasRequestedRecovery = false

    init(
        route: TerminalBulkRoute,
        publishEvent: @escaping PublishEvent
    ) {
        self.route = route
        self.publishEvent = publishEvent
    }

    func beginInitialSnapshot(id: UInt32) async throws {
        initialSnapshotID = id
        isSnapshotting = true
        try await setOutputCredit(0)
    }

    func handle(_ frame: TerminalMultiplexFrame) async throws -> Bool {
        switch frame.opcode {
        case .output:
            try await handleOutput(frame)
        case .snapshotStart:
            do {
                try assembler.start(frame)
                isSnapshotting = true
                hasRequestedRecovery = false
            } catch {
                try await recover()
            }
        case .snapshotChunk:
            do {
                try assembler.append(frame)
            } catch {
                try await recover()
            }
        case .snapshotEnd:
            try await handleSnapshotEnd(frame)
        case .modelRestore:
            try await prepareForModelRestore(frame)
        case .end:
            guard frame.correlationID == 0,
                (try? JSONDecoder().decode(TerminalMultiplexEndRecord.self, from: frame.payload))
                    != nil
            else {
                throw TerminalMultiplexDeliveryError.invalidOutput
            }
            pendingEndSequence = frame.sequence
            await publishEndIfParsed()
        default:
            return false
        }
        return true
    }

    func acknowledgeOutput(endSequence: UInt64, receiverQueueBytes: UInt32) async throws {
        guard endSequence > parsedSequence, endSequence <= expectedSequence else {
            try await recover()
            throw TerminalMultiplexDeliveryError.invalidAcknowledgement
        }
        let acknowledged = endSequence - parsedSequence
        guard acknowledged <= UInt32.max else {
            try await recover()
            throw TerminalMultiplexDeliveryError.invalidAcknowledgement
        }
        parsedSequence = endSequence
        try await sendAck(
            kind: 0,
            correlationID: 0,
            sequence: endSequence,
            acknowledgedBytes: UInt32(acknowledged),
            receiverQueueBytes: receiverQueueBytes
        )
        await publishEndIfParsed()
    }

    func acknowledgeSnapshot(id: UInt32) async throws {
        guard let snapshot = awaitingSnapshot, snapshot.id == id else {
            try await recover()
            throw TerminalMultiplexDeliveryError.invalidAcknowledgement
        }
        awaitingSnapshot = nil
        parsedSequence = snapshot.coverageEndSequence
        expectedSequence = snapshot.coverageEndSequence
        try await sendAck(
            kind: 2,
            correlationID: snapshot.id,
            sequence: snapshot.coverageEndSequence,
            acknowledgedBytes: 0,
            receiverQueueBytes: 0
        )
        isSnapshotting = false
        try await setOutputCredit(1_024 * 1_024)
        let queued = pendingOutput
        pendingOutput.removeAll(keepingCapacity: true)
        for output in queued {
            if output.endSequence > snapshot.pendingDeliveryStartSequence,
                output.endSequence <= snapshot.coverageEndSequence
            {
                continue
            }
            guard output.endSequence > snapshot.pendingDeliveryStartSequence else {
                try await recover()
                throw TerminalMultiplexDeliveryError.invalidOutput
            }
            try await deliver(output)
        }
        if snapshot.id == initialSnapshotID {
            await publishEvent(.subscribed)
        }
    }

    func currentParsedSequence() -> UInt64 {
        parsedSequence
    }

    func applyOutputCredit(_ bytes: UInt32, reason: UInt8) async throws {
        let clamped = bytes == 0 ? 0 : min(8 * 1_024 * 1_024, max(512 * 1_024, bytes))
        try await setOutputCredit(clamped, reason: reason)
    }

    func suspendDelivery() async throws {
        try await setOutputCredit(0)
    }

    func beginReveal() async throws {
        isSnapshotting = true
        hasRequestedRecovery = true
        awaitingSnapshot = nil
        assembler.clear()
        try await setOutputCredit(0)
    }

    private func handleOutput(_ frame: TerminalMultiplexFrame) async throws {
        guard frame.sequence >= UInt64(frame.payload.count) else {
            try await recover()
            throw TerminalMultiplexDeliveryError.invalidOutput
        }
        let output = PendingOutput(
            bytes: frame.payload,
            startSequence: frame.sequence - UInt64(frame.payload.count),
            endSequence: frame.sequence
        )
        if isSnapshotting {
            pendingOutput.append(output)
        } else {
            try await deliver(output)
        }
    }

    private func deliver(_ output: PendingOutput) async throws {
        guard output.endSequence > expectedSequence else { return }
        var bytes = output.bytes
        var startSequence = output.startSequence
        if startSequence < expectedSequence {
            let overlap = Int(expectedSequence - startSequence)
            bytes = bytes.subdata(in: overlap..<bytes.count)
            startSequence = expectedSequence
        }
        guard startSequence == expectedSequence,
            String(data: bytes, encoding: .utf8) != nil
        else {
            try await recover()
            throw TerminalMultiplexDeliveryError.invalidOutput
        }
        expectedSequence = output.endSequence
        await publishEvent(
            .output(TerminalOutputChunk(bytes: bytes, endSequence: output.endSequence))
        )
    }

    private func handleSnapshotEnd(_ frame: TerminalMultiplexFrame) async throws {
        do {
            let snapshot = try assembler.finish(frame)
            awaitingSnapshot = snapshot
            await publishEvent(.snapshot(snapshot))
        } catch TerminalSnapshotAssemblerError.superseded {
            // Why: supersession is normal snapshot arbitration; its replacement stays on this stream.
            return
        } catch TerminalSnapshotAssemblerError.unavailable(let status) {
            throw TerminalMultiplexDeliveryError.snapshotUnavailable(status: status)
        } catch {
            try await recover()
        }
    }

    private func prepareForModelRestore(_ frame: TerminalMultiplexFrame) async throws {
        guard frame.correlationID == 0,
            let record = try? JSONDecoder().decode(
                TerminalMultiplexModelRestoreRecord.self,
                from: frame.payload
            ),
            record.markerSeq == String(frame.sequence)
        else {
            throw TerminalMultiplexDeliveryError.invalidOutput
        }
        isSnapshotting = true
        hasRequestedRecovery = record.snapshotFollows
        awaitingSnapshot = nil
        assembler.clear()
        try await setOutputCredit(0)
    }

    private func recover() async throws {
        guard !hasRequestedRecovery else {
            throw TerminalMultiplexDeliveryError.repeatedRecovery
        }
        hasRequestedRecovery = true
        isSnapshotting = true
        awaitingSnapshot = nil
        assembler.clear()
        try await setOutputCredit(0)
        let correlationID = try await route.allocateCorrelationID()
        let payload = try JSONEncoder().encode(
            TerminalMultiplexSnapshotRequestRecord(
                requestedScrollbackRows: 1_000,
                snapshotMaxBytes: nil
            )
        )
        try await send(
            opcode: .snapshotRequest,
            sequence: parsedSequence,
            correlationID: correlationID,
            payload: payload
        )
    }

    private func setOutputCredit(_ bytes: UInt32, reason: UInt8 = 0) async throws {
        let threshold = bytes == 0 ? 16 * 1_024 : min(256 * 1_024, max(16 * 1_024, bytes / 8))
        let record = TerminalMultiplexCreditRecord(
            direction: 0,
            reason: reason,
            maxInFlightBytes: bytes,
            acknowledgeEveryBytes: threshold,
            maxFrameBytes: 64 * 1_024
        )
        guard let payload = TerminalMultiplexFlowRecordCodec.encode(record) else {
            throw TerminalMultiplexDeliveryError.invalidOutput
        }
        try await send(opcode: .credit, payload: payload)
    }

    private func sendAck(
        kind: UInt8,
        correlationID: UInt32,
        sequence: UInt64,
        acknowledgedBytes: UInt32,
        receiverQueueBytes: UInt32
    ) async throws {
        let record = TerminalMultiplexAckRecord(
            kind: kind,
            status: 0,
            errorCode: 0,
            acknowledgedBytes: acknowledgedBytes,
            cumulativeSequence: sequence,
            receiverQueueBytes: receiverQueueBytes
        )
        guard let payload = TerminalMultiplexFlowRecordCodec.encode(record) else {
            throw TerminalMultiplexDeliveryError.invalidAcknowledgement
        }
        try await send(
            opcode: .ack,
            sequence: sequence,
            correlationID: correlationID,
            payload: payload
        )
    }

    private func send(
        opcode: TerminalMultiplexOpcodeWire,
        sequence: UInt64 = 0,
        correlationID: UInt32 = 0,
        payload: Data = Data()
    ) async throws {
        try await route.send(
            opcode: opcode,
            sequence: sequence,
            correlationID: correlationID,
            payload: payload
        )
    }

    private func publishEndIfParsed() async {
        guard let pendingEndSequence, parsedSequence >= pendingEndSequence else { return }
        self.pendingEndSequence = nil
        await publishEvent(.ended)
    }
}
